src/horde_map.h

```diff
@@ -0,0 +1,257 @@
+                std::unordered_map<tripoint_abs_ms, horde_entity>::iterator inner_iter;
+                int filter = horde_map_flavors::active | horde_map_flavors::idle | horde_map_flavors::dormant;
+            public:
+                friend horde_map;
+                // TODO: ideally these would be private, no use case for constructing them outside of horde_map
+                // No args gets you the end() iterator.
+                explicit iterator() {}
+                explicit iterator( const horde_map &p ) : parent( const_cast<horde_map *>( &p ) ),
+                    outer_map( &parent->active_monster_map ),
+                    outer_iter( outer_map->begin() ) {
+                    insure_valid();
+                }
+                explicit iterator( const horde_map &p, int filt ) : parent( const_cast<horde_map *>( &p ) ),
+                    outer_map( &parent->active_monster_map ),
+                    outer_iter( outer_map->begin() ),
+                    filter( filt ) {
+                    insure_valid();
+                }
+                // Sets the members directly. TODO: make private but accessable to horde_map
+                explicit iterator( const horde_map &p, map_type &m, map_type::iterator oi,
+                                   std::unordered_map<tripoint_abs_ms, horde_entity>::iterator ii ) : parent( const_cast<horde_map *>
+                                               ( &p ) ), outer_map( &m ),
+                    outer_iter( oi ), inner_iter( ii ) {}
+                void next_map() {
+                    if( outer_map == &parent->active_monster_map ) {
+                        if( filter & horde_map_flavors::idle ) {
+                            outer_map = &parent->idle_monster_map;
+                        } else if( filter & horde_map_flavors::dormant ) {
+                            outer_map = &parent->dormant_monster_map;
+                        } else {
+                            outer_map = nullptr;
+                        }
+                    } else if( outer_map == &parent->idle_monster_map ) {
+                        if( filter & horde_map_flavors::dormant ) {
+                            outer_map = &parent->dormant_monster_map;
+                        } else {
+                            outer_map = nullptr;
+                        }
+                    } else {
+                        // At the end.
+                        outer_map = nullptr;
+                    }
+                }
+                void insure_valid() {
+                    // Iterate forward if we don't have a valid initial state.
+                    if( !( filter & horde_map_flavors::active ) || outer_map->empty() ) {
+                        do {
+                            next_map();
+                            if( outer_map == nullptr ) {
+                                return;
+                            }
+                        } while( outer_map->empty() );
```

Comment:
```suggestion
                        } while( outer_map != nullptr && outer_map->empty() );outer_map->empty() );
```

Reply:
This actually does need to be separate because if outer_map equals nullptr here it means the whole container is empty and we want to exit, but if outer_map->empty() == true it just means the current sub-map is empty and a later one might not be, so we try again.

Reply:
I believe my change is a refactor:
1. `outer_map == nullptr`:
   - `while(outer_map != nullptr && outer_map->empty())` :=
   - `while(nullptr != nullptr && outer_map->empty())` :=
   - `while(false && outer_map->empty())` :=
   - `while(false)`
   - ends the cycle
1. `outer_map.empty() is true`:
   - `while(outer_map != nullptr && outer_map->empty())` :=
   - `while(true && true)` :=
   - next cycle
1. `outer_map.empty() is false`:
   - `while(outer_map != nullptr && outer_map->empty())` :=
   - `while(true && false)` :=
   - ends the cycle

Unless I am blundering. The same is true for the original code.

Reply:
I see my blunder now!